### PR TITLE
Webclient redesign

### PR DESCRIPTION
**Pull request by tomas-langer**
WebClient redesign

Resolves:
- #6669 
- #7223 
- #7078 
- #7242 
- #6774 
- #7138 for both HTTP/1 and HTTP/2

Follow up issue: #7269 

Notable changes:
- single API that can run any HTTP version
- upgrade done through HTTP/1.1 client always (implemented both for WebSocket and HTTP/2)
- HTTP client SPI (HTTP/1.1 and HTTP/2 implementations)
- Client SPI (websocket implementation)
- Cookie manager support for HTTP/1.1 and HTTP/2
- WebClientService support for HTTP/1.1 and HTTP/2
- HttpClientResponse.request(Class) now returns a typed entity response, HttpClientResponse.requestEntity(Class) returns only the response entity (and throws if non-success status family)

### CODE REVIEW

File: examples/security/webserver-digest-auth/src/test/java/io/helidon/security/examples/webserver/digest/WebClientAuthenticationService.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public WebClientServiceResponse handle(Chain chain, WebClientServiceRequest requ
         if (challenge == null) {
             return response;
         }
-        String uri = request.uri().path();
+        String uri = request.uri().path().path();
```
Comment (spericas) [inline examples/security/webserver-digest-auth/src/test/java/io/helidon/security/examples/webserver/digest/WebClientAuthenticationService.java:59]: `.path().path()` looks a bit odd. Is this to convert to String?

File: jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java [hunk 2]
```diff
@@ -210,22 +203,10 @@ public ClientResponse apply(ClientRequest request) {
         Http1ClientRequest httpRequest = mapRequest(request);
 
         if (request.hasEntity()) {
-            // if following redirects we need to buffer entity for WebClient
-            if (httpRequest.followRedirects()) {
```
Comment (spericas) [inline jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java:214]: Are you positive this is no longer necessary? Certain redirects will require the entity to be sent again. Maybe this is handled elsewhere.

Comment (spericas) [reply jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java:214]: Are you saying that a developer should know ahead of time that a redirect is possible and provide an in-memory entity for a potential POST/PUT redirect?

File: examples/security/webserver-digest-auth/src/test/java/io/helidon/security/examples/webserver/digest/WebClientAuthenticationService.java (hunk 1 above)
Comment (tomas-langer) [reply examples/security/webserver-digest-auth/src/test/java/io/helidon/security/examples/webserver/digest/WebClientAuthenticationService.java:59]: yes, there are two methods on `UriPath` - `path()` and `rawPath()` to keep it consistent with all other URI components. I do agree it looks odd, we may want to do some renaming of methods here

File: jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java (hunk 2 above)
Comment (tomas-langer) [reply jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java:214]: buffering into memory should at least be limited by a certain amount of bytes. This is really dangerous.
Also this should not be a feature of the connector, but rather of the webclient - as it takes care of redirection itself (and we do not have it yet implemented, but still responsibility of webclient).

File: integrations/vault/auths/common/src/main/java/io/helidon/integrations/vault/auths/common/NoVaultAuth.java [hunk 3]
```diff
@@ -61,7 +61,7 @@ public Optional<RestApi> authenticate(Config config, Vault.Builder vaultBuilder)
                                    .webClientBuilder(webclient -> {
                                        webclient.baseUri(address + "/v1");
                                        vaultBuilder.baseNamespace()
-                                               .ifPresent(ns -> webclient.header(VAULT_NAMESPACE_HEADER_NAME, ns));
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply integrations/vault/auths/common/src/main/java/io/helidon/integrations/vault/auths/common/NoVaultAuth.java:64]: the new code is `.ifPresent(ns -> webclient.addHeader(VAULT_NAMESPACE_HEADER_NAME, ns));`
do not understand your comment at all

File: nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2CallChainBase.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    private ClientResponseHeaders responseHeaders;
+    private Http.Status responseStatus;
+    private Http2ConnectionAttemptResult.Result result;
+
+    Http2CallChainBase(WebClient webClient,
+                       HttpClientConfig clientConfig,
+                       Http2ClientProtocolConfig protocolConfig,
+                       Http2ClientRequestImpl clientRequest,
+                       CompletableFuture<WebClientServiceResponse> whenComplete,
+                       Function<Http1ClientRequest, Http1ClientResponse> http1EntityHandler) {
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/http2/webclient/src/main/java/io/helidon/nima/http2/webclient/Http2CallChainBase.java:66]: I don't understand your comment.

File: nima/http2/webclient/pom.xml [hunk 5]
```diff
@@ -89,8 +94,45 @@
                             <artifactId>helidon-common-features-processor</artifactId>
                             <version>${helidon.version}</version>
                         </path>
+                        <path>
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/http2/webclient/pom.xml:97]: We do not use a Jira.
I think that is a possibility for the future (or some other approach that would require less configuration).

File: nima/tests/integration/http2/client/src/test/java/io/helidon/nima/tests/integration/http2/client/HeadersTest.java [hunk 6]
```diff
@@ -33,21 +43,12 @@
...
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.TimeoutException;
-
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.is;
 
+@Disabled
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/tests/integration/http2/client/src/test/java/io/helidon/nima/tests/integration/http2/client/HeadersTest.java:51]: all methods failed in this test. I will create follow up issues.

File: nima/tests/integration/webclient/webclient/src/test/java/io/helidon/nima/tests/integration/webclient/CookieTest.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import org.junit.jupiter.api.TestMethodOrder;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class CookieTest {
+    private static WebServer server;
+    private static Http1Client client;
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/tests/integration/webclient/webclient/src/test/java/io/helidon/nima/tests/integration/webclient/CookieTest.java:43]: This was brought in from another commit by Santiago, I did not change all usages.

File: nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/ClientResponseTyped.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Http client response explicitly typed.
+ * The typed entity does not extend {@link java.lang.AutoCloseable}, as in most cases the entity is fully read into memory.
+ * Kindly make sure that you fully consume the entity in cases where it cannot be buffered (depends on media support used).
+ *
+ * @param <T> type of the response
+ */
+public interface ClientResponseTyped<T> extends ClientResponseBase {
+    /**
+     * Entity of the requested type.
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/ClientResponseTyped.java:28]: It is the type requested by the user, such as `httpRequest.request(String.class)`.

File: jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java (hunk 2 above)
Comment (spericas) [reply jersey/connector/src/main/java/io/helidon/jersey/connector/HelidonConnector.java:224]: @tomas-langer We can certainly included it in the WebClient and set a size limit. I will create a follow-up issue.

**Review by spericas — APPROVED** [review]
Great improvement over previous API!

File: nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/HttpClientConfigBlueprint.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    Optional<UriFragment> baseFragment();
+
+    /**
+     * Socket options for connections opened by this client.
+     * If there is a value explicitly configured on this type and on the socket options,
+     * the one configured on this type's builder will win:
+     * <ul>
+     *     <li>{@link #readTimeout()}</li>
+     *     <li>{@link #connectTimeout()} eout(java.time.Duration)} (Duration)}</li>
```
Comment (danielkec) [inline nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/HttpClientConfigBlueprint.java:78]: eout?

File: nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/HttpClientConfigSupport.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+         * obtained from a string, in other cases, use an alternative with {@link io.helidon.common.http.Http.HeaderName}
+         * or {@link io.helidon.common.http.Http.HeaderValue}.
+         *
+         * @param builder builder to update
+         * @param name name of the header
+         * @param value value of the header
+         * @see #addHeader(Http.HeaderValue)
+         */
+        @Prototype.BuilderMethod
+        static void addHeader(HttpClientConfig.BuilderBase<?, ?> builder, String name, String value) {
```
Comment (danielkec) [inline nima/webclient/api/src/main/java/io/helidon/nima/webclient/api/HttpClientConfigSupport.java:90]: Javadoc `Set` vs. method name `add`

**Review by danielkec — APPROVED** [review]

**Comment by tomas-langer** [comment]
Updated #7269  with unresolved comments from this PR (except for the inconsistent CR/LF as I cannot find what is wrong)

**Pull request merged**